File: packages/core/src/models/history.rs
// Implementation according to
// https://modelcontextprotocol.io/docs/concepts/sampling/

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub struct HistoryMessage {
    pub role: Role,
    pub content: Vec<Content>,
}

impl HistoryMessage {
    pub fn from_string(role: Role, content: &str) -> Self {
        Self {
            role,
            content: vec![Content::Text {
                content_type: ContentType::Text,
                text: content.to_string(),
            }],
        }
    }
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum Role {
    System,
    User,
    Assistant,
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
#[serde(rename_all = "lowercase")]
#[serde(untagged)]
pub enum Content {
    Text {
        #[serde(rename = "type")]
        content_type: ContentType,
        text: String,
    },
    Image {
        #[serde(rename = "type")]
        content_type: ContentType,
        data: String,
        mime_type: String,
    },
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub enum ContentType {
    Text,
    #[serde(rename = "image_url")]
    ImageUrl,
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
#[serde(untagged)]
pub enum ResponseFormat {
    String(String),
    Object(serde_json::Value),
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
pub struct StreamOptions {
    pub include_usage: bool,
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
pub struct History {
    pub model: String,
    pub messages: Vec<HistoryMessage>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream: Option<bool>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub stream_options: Option<StreamOptions>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_completion_tokens: Option<u32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub top_p: Option<f32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub seed: Option<u32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub presence_penalty: Option<f32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub frequency_penalty: Option<f32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub user: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub stop: Option<Vec<String>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_format: Option<ResponseFormat>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub n: Option<u32>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub tools: Option<Vec<Tool>>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_choice: Option<ToolChoice>,
}

impl History {
    pub fn new(model: String, messages: Vec<HistoryMessage>) -> Self {
        Self {
            model,
            messages,
            stream: Some(true),
            stream_options: None,
            max_completion_tokens: None,
            top_p: None,
            temperature: None,
            seed: None,
            presence_penalty: None,
            frequency_penalty: None,
            user: None,
            stop: None,
            response_format: None,
            n: None,
            tools: None,
            tool_choice: None,
        }
    }

    pub fn push_message(&mut self, message: HistoryMessage) {
        self.messages.push(message);
    }

    pub fn set_system_prompt(&mut self, prompt: String) {
        let system_prompt_index = self
            .messages
            .iter()
            .position(|message| message.role == Role::System);

        if let Some(index) = system_prompt_index {
            self.messages[index].content = vec![Content::Text {
                content_type: ContentType::Text,
                text: prompt,
            }];
            return;
        }

        self.messages.insert(
            0,
            HistoryMessage {
                role: Role::System,
                content: vec![Content::Text {
                    content_type: ContentType::Text,
                    text: prompt,
                }],
            },
        );
    }

    pub fn set_stream(&mut self, stream: bool) {
        self.stream = Some(stream);
    }
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
pub struct Tool {
    #[serde(rename = "type")]
    pub tool_type: ToolType,
    pub function: HistoryFunction,
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
#[serde(rename_all = "lowercase")]
pub enum ToolType {
    Function,
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
pub struct HistoryFunction {
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parameters: Option<serde_json::Value>,
    pub strict: Option<bool>,
}

#[derive(Serialize, Deserialize, JsonSchema, Debug, Clone)]
#[serde(rename_all = "lowercase", untagged)]
pub enum ToolChoice {
    None,
    Auto,
    Required,
    Specific {
        r#type: ToolType,
        function: HistoryFunction,
    },
}

File: packages/ui/lib/schema/llm/history.ts
export interface IHistory {
	frequency_penalty?: number | null;
	max_completion_tokens?: number | null;
	messages: IMessageElement[];
	model: string;
	n?: number | null;
	presence_penalty?: number | null;
	preset?: null | string;
	response_format?: any;
	seed?: number | null;
	stop?: string[] | null;
	stream?: boolean | null;
	stream_options?: null | IStreamOptionsObject;
	temperature?: number | null;
	tool_choice?: null | IToolChoiceObject;
	tools?: IToolElement[] | null;
	top_p?: number | null;
	usage?: null | IUsageObject;
	user?: null | string;
	[property: string]: any;
}

export interface IMessageElement {
	annotations?: IAnnotationElement[] | null;
	content: IContentElement[] | string;
	name?: null | string;
	role: IRole;
	tool_call_id?: null | string;
	tool_calls?: IToolCallElement[] | null;
	[property: string]: any;
}

export interface IAnnotationElement {
	type: string;
	url_citation?: null | IURLCitationObject;
	[property: string]: any;
}

export interface IURLCitationObject {
	content?: null | string;
	end_index: number;
	start_index: number;
	title: string;
	url: string;
	[property: string]: any;
}

export interface IContentElement {
	text?: string;
	type: IContentType;
	image_url?: IImageURL;
	audio_url?: string;
	video_url?: string;
	document_url?: string;
	[property: string]: any;
}

export interface IImageURL {
	detail?: null | string;
	url: string;
	[property: string]: any;
}

export enum IContentType {
	AudioURL = "audio_url",
	DocumentURL = "document_url",
	IImageURL = "image_url",
	Text = "text",
	VideoURL = "video_url",
}

export enum IRole {
	Assistant = "assistant",
	Function = "function",
	System = "system",
	Tool = "tool",
	User = "user",
}

export interface IToolCallElement {
	function: IToolCallFunction;
	id: string;
	type: string;
	[property: string]: any;
}

export interface IToolCallFunction {
	arguments: string;
	name: string;
	[property: string]: any;
}

export interface IStreamOptionsObject {
	include_usage: boolean;
	[property: string]: any;
}

export interface IToolChoiceObject {
	function: IToolChoiceFunction;
	type: IToolChoiceType;
	[property: string]: any;
}

export interface IToolChoiceFunction {
	description?: null | string;
	name: string;
	parameters: IParameters;
	[property: string]: any;
}

export interface IParameters {
	properties?: { [key: string]: IPropertyValue } | null;
	required?: string[] | null;
	type: ITypeEnum;
	[property: string]: any;
}

export interface IPropertyValue {
	description?: null | string;
	enum?: string[] | null;
	items?: null | IPropertyValue;
	properties?: { [key: string]: IPropertyValue } | null;
	required?: string[] | null;
	type?: ITypeEnum | null;
	[property: string]: any;
}

export enum ITypeEnum {
	Array = "array",
	Boolean = "boolean",
	Null = "null",
	Number = "number",
	Object = "object",
	String = "string",
}

export enum IToolChoiceType {
	Function = "function",
}

export interface IToolElement {
	function: IToolChoiceFunction;
	type: IToolChoiceType;
	[property: string]: any;
}

export interface IUsageObject {
	include: boolean;
	[property: string]: any;
}
